Packs/PrismaCloud/TestPlaybooks/playbook-RedLockTest.yml

```diff
@@ -1,14 +1,14 @@
 id: RedLockTest
-version: -1
+version: 34
```

Comment:
```suggestion
version: -1
```

Reply:
Update playbook version.  In regards to the context data, all the keys are in `ContextKey` format except for the items under the `Data` key.  This is the key that contains the raw results of the RQL query.  As this query is freeform the data returned is unstructured.  

For example 
```
!redlock-get-rql-response rql=`config where api.name = 'aws-ec2-describe-instances' as X; config where api.name = 'aws-ec2-describe-security-groups' as Y; filter '$.X.publicIpAddress exists and not $.X.publicIpAddress is empty  and $.X.securityGroups[*].groupName == $.Y.groupName and $.Y.ipPermissions[*].ipRanges[*] contains 0.0.0.0/0 and $.Y.ipPermissions[*].ipProtocol == -1'; show X;` limit=10
```
vs
```
!redlock-get-rql-response rql=`config where api.name = 'aws-ec2-describe-volumes' AND json.rule =  tags[*].key != DataClassification` limit=10
```

return different keys under `Data`.  So it is not possible to map the results to a new hardcoded object with correctly formatted keys.

I am not sure how this can be resolved given the nature of the command being run without recursively iterating through the `Data` key but I did not think recursive functions are supported in integrations at this time?  Last I have checked the `flatten` command will only flatten 1 level not any nested objects so this is also not viable as far as I can tell.

Maybe we can sync on this next week briefly to try and resolve this or if you have an example of where this is done in another integration I can try and take similar logic.  Let me know your thoughts.

Reply:
@nericksen regarding the context outputs, make sure that every output you inject to context in the py file is declared in the yml file, they should be aligned.

regarding the context keys format, we do allow recursive functions, you can try this implementation:
```
def format_response(response):
    if response and isinstance(response, dict):
            response = {pascalToSpace(key).replace(" ", ""): format_response(value) for key, value in response.items()}
    elif response and isinstance(response, list):
        response = [format_response(item) for item in response]
    elif response:
        response = response
    return response
```

Reply:
ohh thats very interesting did not realize we could do recursion in intergations. Thanks! will work to implement this today.

Reply:
Really like that function.  Will use that all the time.  

I think the last piece now is resolving the the context injection to the outputs in the yml.  The standard RQL response keys should be the same at this point.

The one piece here that is incomplete is the `Data` key as it contains an object with variable keys based on the specfic RQL.
```
 93     - contextPath: Redlock.RQL.Response.Data
 94       description: 'The data object returned by the RQL response.'
```

Should I use an example query to populate this? It wouldn't be generic.

Reply:
@nericksen Regarding the `Data` object, if you know all the fields that this kind of object can consist so we should declare them in the yml file. If not, just declare as you did in the comment above, and if possible attach a link to docs or something that the user can understand what does this object represents.

Reply:
Reference added to readme